### PR TITLE
Feat/updated cli

### PR DESCRIPTION
**Pull request by chrisdutz**
# Description

Refaqctoring of the commandline functionality of the DataNode and ConfigNode to use a shared code-base built on top of Apache Commons CLI.

#13193

### CODE REVIEW

**Comment by chrisdutz** [comment]
Will look into the remaining test-failures on Monday.

File: .idea/icon.png
Comment (OneSizeFitsQuorum) [inline .idea/icon.png:1]: Remove this file?

Comment (chrisdutz) [reply .idea/icon.png:1]: I intentionally added it, as it gives IoTDB a nice logo in IntelliJ ... but if people don't want it ... sure.

Comment (OneSizeFitsQuorum) [reply .idea/icon.png:1]: Got it！Great idea！I think people will love it

Comment (chrisdutz) [reply .idea/icon.png:1]: Stumbled over it when looking into flink as part of my "who uses singletons" research ;-) ... now PLC4X has a nice toddy smiling at me in IntelliJ ;-)
Bildschirmfoto 2024-08-20 um 14 08 31

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java [hunk 1]
```diff
@@ -242,6 +254,57 @@ protected void doAddNode() {
...
+              .stream()
+              .map(TDataNodeConfiguration::getLocation)
+              .filter(location -> location.getDataNodeId() == removeNodeId)
+              .findFirst();
+      if (!dataNodeLocationOpt.isPresent()) {
+        throw new IoTDBException("Invalid node-id", -1);
+      }
+      TDataNodeLocation dataNodeLocation = dataNodeLocationOpt.get();
+
+      //
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/DataNode.java:288]: remove this?

**Review by liyuheng55555 — CHANGES_REQUESTED** [review]

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/ServerCommandLine.java [hunk 2]
```diff
@@ -18,50 +18,97 @@
...
+      // As we only support start and remove and one has to be selected,
+      // no need to check if OPTION_REMOVE is set.
+      else {
+        Number nodeId = (Number) cmd.getParsedOptionValue(OPTION_REMOVE);
+        if (nodeId != null) {
+          remove(nodeId.longValue());
+        } else {
+          remove(null);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (chrisdutz) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/ServerCommandLine.java:86]: There is no "no option" ... that's why I even left a comment in the code.

The "command" OptionGroup is mandatory, so one of both has to be selected and not selecting an option results in the usage being printed. This is tested by the ServerCommandLineTest.testNoArgs test.

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Please update this for removeConfigNode/DatNode.sh/bat such as https://github.com/apache/iotdb/blob/master/iotdb-core/datanode/src/assembly/resources/sbin/remove-datanode.sh#L27-L28

**Comment by chrisdutz** [comment]
scripts updated.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/ServerCommandLine.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/ServerCommandLine.java:86]: OK, that make sense.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java [hunk 3]
```diff
@@ -120,7 +124,7 @@ public void run() {
     try (ConfigNodeClient client =
         ConfigNodeClientManager.getInstance().borrowClient(ConfigNodeInfo.CONFIG_REGION_ID)) {
       isReportSuccess =
-          client.reportDataNodeShutdown(DataNode.generateDataNodeLocation()).getCode()
+          client.reportDataNodeShutdown(nodeLocation).getCode()
```
Comment (liyuheng55555) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:127]: What's the difference between "call DataNode.generateDataNodeLocation() here" and "pass the nodeLocation in when constructing the IoTDBShutdownHook" ?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java [hunk 4]
```diff
@@ -49,8 +50,11 @@ public class IoTDBShutdownHook extends Thread {
 
   private static final Logger logger = LoggerFactory.getLogger(IoTDBShutdownHook.class);
 
-  public IoTDBShutdownHook() {
+  private final TDataNodeLocation nodeLocation;
+
+  public IoTDBShutdownHook(TDataNodeLocation nodeLocation) {
```
Comment (liyuheng55555) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:55]: How about rename this class to DataNodeShutDownHook? @OneSizeFitsQuorum what do you think

**Review by liyuheng55555 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java (hunk 3 above)
Comment (chrisdutz) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:127]: Less singleton pattern?

**Comment by OneSizeFitsQuorum** [comment]
Hi, chris, the ainode has been [merged](https://lists.apache.org/thread/rk4mww77nwmln7htdjzvsm4v4xtv0sts) into master, maybe we can modify ainode script in this pr?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java (hunk 4 above)
Comment (OneSizeFitsQuorum) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:55]: This PR is irrelevant  with here. We can do it in next PR.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java (hunk 3 above)
Comment (liyuheng55555) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:127]: It's fine, just ask for curious.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java (hunk 4 above)
Comment (liyuheng55555) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/service/IoTDBShutdownHook.java:55]: lgtm

**Review by liyuheng55555 — APPROVED** [review]
Good work! The code here is much clean now.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

Thanks a lot for working on this!

**Pull request merged**